core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+		writer.write_u8(self.log_size)?;
+		writer.write_u64(self.idx)
+	}
+}
+
+/// Segment of a PMMR: unpruned leaves and the necessary data to verify
+/// segment membership in the original MMR.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Segment<T> {
+	identifier: SegmentIdentifier,
+	hashes: Vec<(u64, Hash)>,
+	leaf_data: Vec<(u64, T)>,
+	proof: SegmentProof,
+}
+
+impl<T> Segment<T> {
+	/// Maximum number of leaves in a segment, given by `2**b`
+	#[inline]
+	fn segment_capacity(&self) -> u64 {
+		1 << self.identifier.log_size
+	}
+
+	/// Offset (in leaf idx) of first leaf in the segment
+	#[inline]
+	fn leaf_offset(&self) -> u64 {
+		self.identifier.idx * self.segment_capacity()
+	}
+
+	/// Number of leaves in this segment. Equal to capacity except for the final segment, which can be smaller
+	#[inline]
+	fn segment_unpruned_size(&self, last_pos: u64) -> u64 {
+		min(
+			self.segment_capacity(),
+			pmmr::n_leaves(last_pos).saturating_sub(self.leaf_offset()),
+		)
+	}
+
+	/// Whether the segment is full (size == capacity)
+	#[inline]
+	fn full_segment(&self, last_pos: u64) -> bool {
+		self.segment_unpruned_size(last_pos) == self.segment_capacity()
+	}
+
+	/// Inclusive range of MMR positions for this segment
+	#[inline]
+	fn segment_pos_range(&self, last_pos: u64) -> (u64, u64) {
+		let segment_size = self.segment_unpruned_size(last_pos);
+		let leaf_offset = self.leaf_offset();
+		let first = pmmr::insertion_to_pmmr_index(leaf_offset + 1);
+		let last = if self.full_segment(last_pos) {
+			pmmr::insertion_to_pmmr_index(leaf_offset + segment_size)
+				+ (self.identifier.log_size as u64)
+		} else {
+			last_pos
+		};
+
+		(first, last)
+	}
+
+	fn get_hash(&self, pos: u64) -> Result<Hash, SegmentError> {
+		self.hashes
+			.binary_search_by(|&(p, _)| p.cmp(&pos))
+			.map(|p| self.hashes[p].1)
+			.map_err(|_| SegmentError::MissingHash(pos))
+	}
+}
+
+impl<T> Segment<T>
+where
+	T: Readable + Writeable + Debug,
+{
+	/// Generate a segment from a PMMR
+	pub fn from_pmmr<U, B>(
+		segment_id: SegmentIdentifier,
+		pmmr: &ReadonlyPMMR<'_, U, B>,
+		prunable: bool,
+	) -> Result<Self, SegmentError>
+	where
+		U: PMMRable<E = T>,
+		B: Backend<U>,
+	{
+		let mut segment = Segment {
+			identifier: segment_id,
+			hashes: Vec::new(),
+			leaf_data: Vec::new(),
+			proof: SegmentProof::empty(),
+		};
+
+		let last_pos = pmmr.unpruned_size();
+		if segment.segment_unpruned_size(last_pos) == 0 {
+			return Err(SegmentError::NonExistent);
+		}
+
+		// Fill leaf data and hashes
+		let (segment_first_pos, segment_last_pos) = segment.segment_pos_range(last_pos);
+		for pos in segment_first_pos..=segment_last_pos {
+			if pmmr::is_leaf(pos) {
+				if let Some(data) = pmmr.get_data_from_file(pos) {
+					segment.leaf_data.push((pos, data));
+					continue;
+				} else if !prunable {
+					return Err(SegmentError::MissingLeaf(pos));
+				}
+			}
+			// TODO: optimize, no need to send every intermediary hash
```

Comment:
I think the list of hashes is the same as the list of pos that return true for `is_prune_root()`. i.e. We need the hash for the root of every pruned subtree.

Reply:
Ok, I think we should leave this for a future PR where we can check it properly, as there are already a lot of moving parts in this PR.